File: android/sdk/ssdid-core/src/main/java/my/ssdid/sdk/api/FlowsApi.kt
package my.ssdid.sdk.api

import my.ssdid.sdk.api.model.AuthSession
import my.ssdid.sdk.api.model.TransactionResult
import my.ssdid.sdk.domain.SsdidClient
import my.ssdid.sdk.domain.model.Identity
import my.ssdid.sdk.domain.model.VerifiableCredential

class FlowsApi internal constructor(private val client: SsdidClient) {
    suspend fun registerWithService(identity: Identity, serviceUrl: String): Result<VerifiableCredential> =
        client.registerWithService(identity, serviceUrl)

    suspend fun authenticate(credential: VerifiableCredential, serviceUrl: String): Result<AuthSession> =
        client.authenticate(credential, serviceUrl).map { resp ->
            AuthSession(sessionToken = resp.session_token, serverDid = resp.server_did)
        }

    suspend fun fetchTransactionDetails(sessionToken: String, serverUrl: String): Result<Map<String, String>> =
        client.fetchTransactionDetails(sessionToken, serverUrl)

    suspend fun signTransaction(
        sessionToken: String,
        identity: Identity,
        transaction: Map<String, String>,
        serverUrl: String
    ): Result<TransactionResult> =
        client.signTransaction(sessionToken, identity, transaction, serverUrl).map { resp ->
            TransactionResult(transactionId = resp.transaction_id, status = resp.status)
        }
}

File: sdk/ios/SsdidCore/Sources/SsdidCore/Api/FlowsApi.swift
import Foundation

/// Facade for SSDID protocol flows (registration, authentication, transaction signing).
public struct FlowsApi {
    let client: SsdidClient

    public func registerWithService(identity: Identity, serviceUrl: String) async throws -> VerifiableCredential {
        try await client.registerWithService(identity: identity, serverUrl: serviceUrl)
    }

    public func authenticate(credential: VerifiableCredential, serviceUrl: String) async throws -> AuthenticateResponse {
        try await client.authenticate(credential: credential, serverUrl: serviceUrl)
    }

    public func fetchTransactionDetails(sessionToken: String, serverUrl: String) async throws -> [String: String] {
        try await client.fetchTransactionDetails(sessionToken: sessionToken, serverUrl: serverUrl)
    }

    public func signTransaction(
        sessionToken: String,
        identity: Identity,
        transaction: [String: String],
        serverUrl: String
    ) async throws -> TxSubmitResponse {
        try await client.signTransaction(
            sessionToken: sessionToken,
            identity: identity,
            transaction: transaction,
            serverUrl: serverUrl
        )
    }
}
